src/join-8-auth.sol

```diff
@@ -0,0 +1,108 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.6.12;
+
+interface VatLike {
+    function slip(bytes32, address, int256) external;
+}
+
+interface GemLike {
+    function decimals() external view returns (uint8);
+    function transfer(address,uint256) external returns (bool);
+    function transferFrom(address,address,uint256) external returns (bool);
+    function erc20Impl() external view returns (address);
+}
+
+// AuthGemJoin8
+// For a token that has a lower precision than 18, has decimals and it is upgradable (like GUSD)
+
+contract AuthGemJoin8 {
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    VatLike public immutable vat;
+    bytes32 public immutable ilk;
+    GemLike public immutable gem;
+    uint256 public immutable dec;
+    uint256 public live;  // Access Flag
+
+    // Events
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Join(address indexed usr, uint256 wad, address indexed msgSender);
+    event Exit(address indexed usr, uint256 wad);
+    event Cage();
+
+    mapping (address => uint256) public implementations;
+
+    constructor(address vat_, bytes32 ilk_, address gem_) public {
+        gem = GemLike(gem_);
+        dec = GemLike(gem_).decimals();
+        require(GemLike(gem_).decimals() < 18, "AuthGemJoin8/decimals-18-or-higher");
+        wards[msg.sender] = 1;
+        live = 1;
+        setImplementation(GemLike(gem_).erc20Impl(), 1);
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        emit Rely(msg.sender);
+    }
+
+    function cage() external auth {
+        live = 0;
+        emit Cage();
+    }
+
+    function setImplementation(address implementation, uint256 permitted) public auth {
+        implementations[implementation] = permitted;  // 1 live, 0 disable
+    }
+
+    function mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "AuthGemJoin8/overflow");
+    }
+
+    function join(address usr, uint256 amt, address msgSender) external auth {
+        require(live == 1, "AuthGemJoin8/not-live");
+        uint256 wad = mul(amt, 10 ** (18 - dec));
+        require(int256(wad) >= 0, "AuthGemJoin8/overflow");
+        require(implementations[gem.erc20Impl()] == 1, "AuthGemJoin8/implementation-invalid");
+        vat.slip(ilk, usr, int256(wad));
+        require(gem.transferFrom(msg.sender, address(this), amt), "AuthGemJoin8/failed-transfer");
```

Comment:
Checking the `GUSD` join8, I think here should be `wad` instead of `amt`.
Diffcheck (`join`):
https://www.diffchecker.com/cYH5XQrE

Reply:
actually I think is correct from the decimals point of view, but the naming between the two joins it's a bit confusing and should be defined for consistency.

Reply:
Renamed `usr` to `urn` in the `join`, but the `amt` / `wad` in join-8 is a better naming scheme imo. If anything a new PR should be opened to fix the join-5.

Reply:
we should also fix the join-5-auth as well.

Reply:
I think `urn` should be adopted also in `exit` if we want to go in that direction.

Reply:
#3